client/pluginmanager/drivermanager/instance.go

```diff
@@ -0,0 +1,443 @@
+// for the context used in the RPC. This cancel func should always be called
+// when the caller is finished with the channel.
+func (i *instanceManager) dispenseFingerprintCh() (<-chan *drivers.Fingerprint, context.CancelFunc, error) {
+	driver, err := i.dispense()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ctx, cancel := context.WithCancel(i.ctx)
+	fingerCh, err := driver.Fingerprint(ctx)
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+
+	return fingerCh, cancel, nil
+}
+
+// fingerprint is the main loop for fingerprinting.
+func (i *instanceManager) fingerprint() {
+	fpChan, cancel, err := i.dispenseFingerprintCh()
+	if err != nil {
+		i.logger.Error("failed to dispense driver plugin", "error", err)
+	}
+
+	// backoff and retry used if the RPC is closed by the other end
+	var backoff time.Duration
+	var retry int
+	for {
+		time.Sleep(backoff)
+		select {
+		case <-i.ctx.Done():
+			cancel()
+			return
+		case fp, ok := <-fpChan:
+			if ok {
+				if fp.Err == nil {
+					i.handleFingerprint(fp)
+				} else {
+					i.logger.Warn("received fingerprint error from driver", "error", fp.Err)
+					i.handleFingerprintError()
+				}
+				continue
+			}
+
+			// if the channel is closed attempt to open a new one
+			newFpChan, newCancel, err := i.dispenseFingerprintCh()
+			if err != nil {
+				i.logger.Warn("failed to fingerprint driver, retrying in 30s", "error", err, "retry", retry)
+				i.handleFingerprintError()
+
+				// Calculate the new backoff
+				backoff = (1 << (2 * uint64(retry))) * driverFPBackoffBaseline
+				if backoff > driverFPBackoffLimit {
+					backoff = driverFPBackoffLimit
+				}
+				// Increment retry counter
+				retry++
+				continue
+			}
+			cancel()
+			fpChan = newFpChan
+			cancel = newCancel
+
+			// Reset backoff
+			backoff = 0
+			retry = 0
+		}
+	}
+}
+
+// handleFingerprintError is called when an error occurred while fingerprinting
+// and will set the driver to unhealthy
+func (i *instanceManager) handleFingerprintError() {
+	di := &structs.DriverInfo{
+		Healthy:           false,
+		HealthDescription: "failed to fingerprint driver",
+		UpdateTime:        time.Now(),
+	}
+	i.updateNodeFromDriver(i.id.Name, di)
+}
+
+// handleFingerprint updates the node with the current fingerprint status
+func (i *instanceManager) handleFingerprint(fp *drivers.Fingerprint) {
+	di := &structs.DriverInfo{
+		Attributes:        fp.Attributes,
+		Detected:          fp.Health != drivers.HealthStateUndetected,
+		Healthy:           fp.Health == drivers.HealthStateHealthy,
+		HealthDescription: fp.HealthDescription,
+		UpdateTime:        time.Now(),
+	}
+	i.updateNodeFromDriver(i.id.Name, di)
+
+	// log detected/undetected state changes after the initial fingerprint
+	if i.hasFingerprinted {
+		if i.lastHealthState != fp.Health {
+			i.logger.Info("driver health state has changed", "previous", i.lastHealthState, "current", fp.Health, "description", fp.HealthDescription)
+		}
+	}
+	i.lastHealthState = fp.Health
```

Comment:
This appears to be written in this goroutine and read from other goroutines, so it needs locking (maybe getter/setter funcs).